src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -124,6 +114,11 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
             {
                 // TODO: ISSUE 31309: Optimize HPACK static table decoding
 
+                if (_state != StreamState.ExpectingHeaders)
```

Comment:
Can you document in a comment when SyncObject is needed / what state it protects?

Reply:
Actually, I think I'm going to just always ensure we take the SyncObject lock, here and elsewhere.

The reason for not locking here in the first place was that we shouldn't have returned from SendAsync yet, and thus there should be no possibility for concurrent access between user and Http2Connection here. But I'm not sure that's always and forever true, e.g. cancellation or other exception. The cost of locking is small, so it seems best to just do it consistently here.